### ISSUE TITLE
Installation

### ISSUE THREAD

**Issue by hannorein** [carlislewishard/swiftest#1]
(I'm starting to work on #openjournals/joss-reviews/issues/5409.)

I'm trying to install the code on a vanilla linux machine without using root access to install system wide package because I think this is similar to what the average user would do. But after trying for an hour or so, I'm still stuck at the installation. Basically, I'm lost in the chain of dependencies. Just trying to install NetCDF requires all kind of other libraries. I think I've managed to get NetCDF working in the end, but now I'm stuck at NetCDF-fortran with a random error message I don't understand. 

I don't want to you to help me install some 3rd party library. But related to this project and the review process, I'm wondering if you can either
1) make the dependencies optional, especially the NetCDF ones, or
2) provide a container that lets me run the code without having to install all the dependencies manually.

**Comment by profminton** [comment]
Hi Hanno. Sorry for the installation woes! Unfortunately, NetCDF is pretty firmly embedded in the project. The library dependency issue has always been a known speedbump with this project that was masked by the fact that Purdue's computer systems were set up by climate modelers with NetCDF stickers on their laptops. I was already considering implementing a Docker container for this, so I'll fast track that now.

**Comment by hannorein** [comment]
I've also tried installing it on the supercomputing cluster which helps with the dependencies. But I'm still getting error messages that I don't know how to resolve. So I'm a bit lost on how to proceed.
```
/home/rein/swiftest/swiftest/src/globals/globals_module.f90:108:47:

  108 |    character(*), parameter :: SWIFTEST_LOG_FILE = "swiftest.log" !! Name of file to use to log output when using "COMPACT" or "PROGRESS" display style
      |                                               1
Error: Symbol ‘swiftest_log_file’ at (1) already has basic type of CHARACTER
```

**Comment by profminton** [comment]
Can you tell me the compiler you are using?

**Comment by hannorein** [comment]
GNU 9.3.0

**Comment by profminton** [comment]
Thank you. I just uncovered a bug I hadn't caught before when compiling using gcc 10.2.0. Also, in the time since we originally submitted to JOSS we uncovered some bugs and also added experimental support for the use of coarray test particles (which is currently turned off by default by CMAKE using preprocessor conditional compilation). The master branch is currently protected until a PR is done. Would you mind switching to the debug branch to test fixes? We have GNU 9.3.0 on our cluster, so I will troubleshoot next.

**Comment by hannorein** [comment]
I get the same error. Do I need to clean some of the build files first? If so , how?

**Comment by profminton** [comment]
Typically I call this sequence of command from the main project directory:
` cmake -P distclean.cmake;mkdir build; cd build;cmake .. -DCMAKE_BUILD_TYPE=release;make`

**Comment by profminton** [comment]
FYI I was able to successfully compile and run a simulation when compiling with GNU 9.3.0

**Comment by hannorein** [comment]
Same error here.

**Comment by profminton** [comment]
Hm. Very strange. I'll keep poking around and see if I can reproduce it or at least identify the root cause. The error appears to be some sort of variable name collision.

**Comment by profminton** [comment]
What version of CMAKE do you have?

**Comment by profminton** [comment]
I don't have a fix yet, but I have a hypothesis. This error would occur if preprocessor directives were ignored.

**Comment by profminton** [comment]
Can you make the project with the following command: 
`make VERBOSE=1` 
And copy the output?

**Comment by hannorein** [comment]
cmake version 3.16.5

**Comment by hannorein** [comment]
```
-bash-4.2$ make VERBOSE=1
/cvmfs/soft.computecanada.ca/gentoo/2020/usr/bin/cmake -S/home/rein/swiftest/swiftest -B/home/rein/swiftest/swiftest/build --check-build-system CMakeFiles/Makefile.cmake 0
/cvmfs/soft.computecanada.ca/gentoo/2020/usr/bin/cmake -E cmake_progress_start /home/rein/swiftest/swiftest/build/CMakeFiles /home/rein/swiftest/swiftest/build/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/home/rein/swiftest/swiftest/build'
make -f ../bin/CMakeFiles/swiftest_driver.dir/build.make ../bin/CMakeFiles/swiftest_driver.dir/depend
make[2]: Entering directory '/home/rein/swiftest/swiftest/build'
cd /home/rein/swiftest/swiftest/build && /cvmfs/soft.computecanada.ca/gentoo/2020/usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/rein/swiftest/swiftest /home/rein/swiftest/swiftest/src /home/rein/swiftest/swiftest/build /home/rein/swiftest/swiftest/bin /home/rein/swiftest/swiftest/bin/CMakeFiles/swiftest_driver.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/rein/swiftest/swiftest/build'
make -f ../bin/CMakeFiles/swiftest_driver.dir/build.make ../bin/CMakeFiles/swiftest_driver.dir/build
make[2]: Entering directory '/home/rein/swiftest/swiftest/build'
[  1%] Building Fortran object ../bin/CMakeFiles/swiftest_driver.dir/globals/globals_module.f90.o
cd /home/rein/swiftest/swiftest/bin && /cvmfs/soft.computecanada.ca/custom/bin/f95  -I/include  -fno-underscoring -march=native -O3 -DNDEBUG -O3 -unroll -finline-functions -ffree-line-length-none -J../include   -fopenmp -c /home/rein/swiftest/swiftest/src/globals/globals_module.f90 -o CMakeFiles/swiftest_driver.dir/globals/globals_module.f90.o
f951: Warning: Nonexistent include directory ‘/include’ [-Wmissing-include-dirs]
/home/rein/swiftest/swiftest/src/globals/globals_module.f90:105:2:

  105 | #ifdef COARRAY
      |  1
Warning: Illegal preprocessor directive
/home/rein/swiftest/swiftest/src/globals/globals_module.f90:107:2:

  107 | #else
      |  1
Warning: Illegal preprocessor directive
/home/rein/swiftest/swiftest/src/globals/globals_module.f90:109:2:

  109 | #endif
      |  1
Warning: Illegal preprocessor directive
/home/rein/swiftest/swiftest/src/globals/globals_module.f90:108:47:

  108 |    character(*), parameter :: SWIFTEST_LOG_FILE = "swiftest.log" !! Name of file to use to log output when using "COMPACT" or "PROGRESS" display style
      |                                               1
Error: Symbol ‘swiftest_log_file’ at (1) already has basic type of CHARACTER
make[2]: *** [../bin/CMakeFiles/swiftest_driver.dir/build.make:63: ../bin/CMakeFiles/swiftest_driver.dir/globals/globals_module.f90.o] Error 1
make[2]: Leaving directory '/home/rein/swiftest/swiftest/build'
make[1]: *** [CMakeFiles/Makefile2:122: ../bin/CMakeFiles/swiftest_driver.dir/all] Error 2
make[1]: Leaving directory '/home/rein/swiftest/swiftest/build'
make: *** [Makefile:130: all] Error 2
```

**Comment by profminton** [comment]
It looks to me like it's not setting any of the flags or getting any of the include directories set. Compare with the equivalent output on my machine. My guess it must be something related to CMAKE not parsing properly. I have CMAKE v.3.22.1. I'll see if I can roll it back to an earlier version and see if I can reproduce the issue.
 
```
daminton at bell-b003 in ~/git_joss/swiftest/build (debug●) 
$ make VERBOSE=1
/depot/daminton/apps/minton_lab_packages/bin/cmake -S/home/daminton/git_joss/swiftest -B/home/daminton/git_joss/swiftest/build --check-build-system CMakeFiles/Makefile.cmake 0
/depot/daminton/apps/minton_lab_packages/bin/cmake -E cmake_progress_start /home/daminton/git_joss/swiftest/build/CMakeFiles /home/daminton/git_joss/swiftest/build//CMakeFiles/progress.marks
make  -f CMakeFiles/Makefile2 all
make[1]: Entering directory `/home/daminton/git_joss/swiftest/build'
make  -f ../bin/CMakeFiles/swiftest_driver.dir/build.make ../bin/CMakeFiles/swiftest_driver.dir/depend
make[2]: Entering directory `/home/daminton/git_joss/swiftest/build'
cd /home/daminton/git_joss/swiftest/build && /depot/daminton/apps/minton_lab_packages/bin/cmake -E cmake_depends "Unix Makefiles" /home/daminton/git_joss/swiftest /home/daminton/git_joss/swiftest/src /home/daminton/git_joss/swiftest/build /home/daminton/git_joss/swiftest/bin /home/daminton/git_joss/swiftest/bin/CMakeFiles/swiftest_driver.dir/DependInfo.cmake --color=
make[2]: Leaving directory `/home/daminton/git_joss/swiftest/build'
make  -f ../bin/CMakeFiles/swiftest_driver.dir/build.make ../bin/CMakeFiles/swiftest_driver.dir/build
make[2]: Entering directory `/home/daminton/git_joss/swiftest/build'
[  1%] Building Fortran object ../bin/CMakeFiles/swiftest_driver.dir/globals/globals_module.f90.o
cd /home/daminton/git_joss/swiftest/bin && /apps/spack/bell/apps/openmpi/4.1.3-gcc-9.3.0-3gfikd5/bin/mpif90  -I/apps/spack/bell/apps/netcdf-fortran/4.5.3-gcc-9.3.0-sdxad2m/include -I/apps/spack/bell/apps/netcdf/4.7.4-gcc-9.3.0-sluqfyd/include -fno-underscoring -march=native -O3 -DNDEBUG -O3 -unroll -finline-functions -ffree-line-length-none -J../include -fopenmp -cpp -c /home/daminton/git_joss/swiftest/src/globals/globals_module.f90 -o CMakeFiles/swiftest_driver.dir/globals/globals_module.f90.o
/depot/daminton/apps/minton_lab_packages/bin/cmake -E cmake_copy_f90_mod ../include/globals.mod ../bin/CMakeFiles/swiftest_driver.dir/globals.mod.stamp GNU
```

**Comment by profminton** [comment]
I may switch gears and try to get a container implemented instead of futzing with versioning incompatibilities. On our supercomputer system, they support Singularity over Docker. Would you be able to run a Singularity container if I was able to create it?

**Comment by hannorein** [comment]
I haven't done it before, but I think so. But maybe wait before you put in a lot of work until the other reviewers chime in. Maybe they have no issues with the installation and I'm the odd one out.

**Comment by profminton** [comment]
I would be surprised if that were so, honestly! I was visiting MSU the week before DDA and got to talking with one of their HPC people about containers, and it made me want to dive in and try it out, because compiling someone else's scientific code seems to always lead to these kinds of headaches (except REBOUND, it seems. That one is magic).

**Comment by hannorein** [comment]
(Only because there are no dependencies!)

**Comment by profminton** [comment]
Hi Hanno, I wanted to give you an update on containerizing the swiftest_driver executable. The most straightforward solution I found was to create a 2-stage Docker build and push the resulting image up to a pubic DockerHub. 

In stage 1 it downloads the complete toolchain needed to compile swiftest_driver and all its dependencies from scratch (I use the Intel HPC compiler package and pull down the netcdf-c and netcdf-fortran source from git and compile those all together). 

In stage 2, it takes the executable and libraries generated in stage 1 and downloads minimal set of packages needed at runtime, the biggest of which are the Intel runtime libraries, but I've always found better performance using ifort over gfortran, so I find it's worth it. The image built at stage 2 that contains only the executable and runtime libraries is what is up on DockerHub. Singularity can pull this off of DockerHub and create a .sif file. 

I adjusted the simulation_class.py script to accept a new argument, `container`, which can be either "docker", "singularity," or None (the default). The main purpose of this is so that if you were to run swiftest from Python, it uses a shell script in place of the swiftest_driver executable when .run() is called.

I'll work on documenting all of this more fully tomorrow, but so far I'm able to get it to work on our HPC cluster computers relatively painlessly.  The main thing that has to happen is that you have to pull the Docker or Singularity images down from DockerHub once before being able to run the driver. To help ease the the workflow, I created a pair of scripts: `docker/pull` and `singularity/pull` as well as a pair of "replacement" executable scripts `docker/bin/swiftest_driver` and `singularity/bin/swiftest_driver`.

On our HPC system, we have to use Singularity for security reasons. The workflow for me looks like this:
```
$ cd git/swiftest/singularity
$ ./pull
[downloads ~3 GB of blob files and generates the swiftest.sif file]
$ source ./setenv.sh
```
Verify that the environment variable SWIFTEST_SIF correctly points to the path of file `swiftest.sif`, otherwise set it yourself.
```$ echo $SWIFTEST_SIF```
Then you can run the singularity replacement script exactly the way you would run the compiled binary executable. 

The main difference is that I had to turn off some of the vector optimizations in the container version of the driver because it needs to work on any x86_64 processor. This means that a native compiled version of the driver may be faster than the containerized version, but I haven't had time to benchmark it yet.

**Comment by profminton** [comment]
I have made some further refinements to the container version of the swiftest_driver executable, including adding a GNU/gfortran version that is much smaller (but also slower) than the Intel version. It still requires some amount of futzing from the user to make it work. I am currently investigating the possibility of packaging up everything into a conda package, including the complete Python dependencies and Fortran build steps, so that one could install everything in a single conda install command.

For now, I added new documentation to the README.md on the debug branch explaining how to use the containerized driver executable. It's a bit different than what I posted last time, so I am including it here: 

**Download the `swiftest_driver` as a Docker or Singularity container.**

The Swiftest driver is available as a Docker container on DockerHub in two versions: Intel and GNU. The Intel version was compiled for the x86_64 CPU using the Intel classic Fortran compiler. The GNU version was compliled for the x86_64 CPU using gfortran. The Intel version is faster than the GNU version (though not as fast as a native compile to the target CPU that you wish to run it on due to vectorization optimizations that Swiftest takes advantage of), however it is much larger: The Intel version is ~2.7GB while the GNU version is ~300MB. The Singularity container pulls from the same DockerHub container.

To facilitate installation of the container, we provide a set of shell scripts to help automate the process of installing container versions of the executable. To install the default Intel version of the docker container from within the `swiftest\` project directory

```
$ cd docker
$ . ./install.sh
```

To install the GNU version:

```
$ cd docker
$ . ./install.sh gnu
```

The Singularity versions are installed the same way, just replace `cd docker` with `cd singularity` above. 

Whether installing either the Docker or Singularity containers, the install script will copy an executable shell script `swiftest_driver` into `swiftest/bin/`. Not that when installing the Singularity container, the install script will set an environment variable called `SWIFTEST_SIF` that must point to the aboslute path of the container file called `swiftest_driver.sif`. To use the driver script in a future shell, rather than running the install script again, we suggest adding the environment variable definition to your shell startup script (e.g. add `export SWIFTEST_SIF="/path/to/swiftest/singularity/swiftest.sif"` to your `.zshrc`)

**Comment by hannorein** [comment]
Hi David!

Thanks for looking into the containerization. I've spent a few more hours trying to get it to work. However, I'm afraid I'm still a bit of a bumpy road as I continue to run into snags along the way...

- It looks like I can't run it on my personal computer (it doesn't have a x86_64 processor)?
- It looks like singularity is deprecated due to security reasons on our cluster. It still works for now, but [we're being asked to move to apptainer](https://docs.alliancecan.ca/wiki/Singularity#Please_use_Apptainer_instead). Looks like mostly a name change?
- The container building seems to have worked! This is probably a really stupid question, but how do I run something now? 
- I've tried to install the python part. Your instructions are for anaconda but [we're being asked not to use](https://docs.alliancecan.ca/wiki/Anaconda/en) anaconda on the cluster. 
- When installing the packages into a virtual environment, I got `ERROR: Package 'xarray' requires a different Python: 3.7.7 not in '>=3.9'`. So I've updated my python version and now it seems to install packages. However, it's installing A LOT of packages. It's been at least half an hour by now. Maybe the cluster login nodes are just really slow right now. I'll let it run and continue when it's done...

**Comment by profminton** [comment]
Thanks for this. Despite all the headaches, this is valuable feedback for us. We've only really had experience working in the Purdue RCAC cluster environment, so it's helpful to see how things break down when the code leaves its natural habitat. 

- What architecture is your PC? I disabled the vectorization optimizations for the containerized version, but there may be other issues that I could troubleshoot using a VM.
- Yes, as far as I can tell it's a name change. It doesn't look like Purdue's RCAC clusters have transitioned to Apptainer yet, but it looks like it should be just a name change.
- I only containerized the compiled executable, `swiftest_driver`, which can be run from the command line and works very similarly to the older Swifter drivers (e.g. `swifter_whm`, `swifter_symba`, etc.). To run it requires a `param.in` file and a set of initial conditions files. Our intention was that for most use cases, you would use the Python package to generate these rather than creating them manually as in the older Swift and Swifter, which brings me to...
- Anaconda. The cause and solution to so many problems. The good people at Purdue RCAC developed a tool that manages conda environments using the Modules system, which allows us to use Anaconda on our clusters, so I've grown accustomed to setting everything up in conda rather then virtual environments and pip. I can write alternative documentation that doesn't rely on conda.
- I've had a crash course in containerization over the past week as I worked on putting together the container for the executable driver. I'm finding that Docker is a good tool for testing the process of getting a set of packages together with all of the dependencies starting from a clean Linux distro. What I can do next is put together a container of the complete Swiftest project (Python + Fortran) that contains absolutely everything required so that you don't have to futz around with package dependencies yourself.
